pkg/aggregator/demultiplexer.go

```diff
@@ -684,6 +673,56 @@ func (d *AgentDemultiplexer) GetMetricSamplePool() *metrics.MetricSamplePool {
 	return d.statsd.metricSamplePool
 }
 
+// GetRoutinesDistribution returns how many routines should be spawned for the
+// DogStatsD pipeline and how many DogStatsD workers should be running.
+func GetRoutinesDistribution() (int, int) {
+	return getRoutinesDistribution(ddruntime.NumVCPU())
+}
+
+func getRoutinesDistribution(vCPUs int) (int, int) {
+	var dsdWorkerCount int
+	var pipelineCount int
+	autoAdjust := config.Datadog.GetBool("dogstatsd_pipeline_autoadjust")
+
+	// no auto-adjust of the pipeline count:
+	// we use the pipeline count configuration
+	// to determine how many workers should be running
+	// ------------------------------------
+
+	if !autoAdjust {
+		pipelineCount = config.Datadog.GetInt("dogstatsd_pipeline_count")
+
+		// - a core for the listener goroutine
+		// - one per aggregation pipeline (time sampler)
+		// - the rest for workers
+		// But we want at minimum 2 workers.
+		dsdWorkerCount = vCPUs - 1 - pipelineCount
+
+		if dsdWorkerCount < 2 {
+			dsdWorkerCount = 2
+		}
+
+		return dsdWorkerCount, pipelineCount
+	}
+
+	// we will auto-adjust the pipeline and workers count
+	// --------------------------------------------------
+
+	dsdWorkerCount = vCPUs / 2
+
+	if dsdWorkerCount < 2 {
+		dsdWorkerCount = 2
+	}
+
+	pipelineCount = dsdWorkerCount - 1
```

Comment:
Could you add a comment here to briefly explain why you're using this heuristic?

For example, are you doing `pipelineCount = dsdWorkerCount - 1` because you want to keep a core for the listener goroutine, and in general you want to have approximately one additional pipeline for every additional dsd worker above 2 workers? (and if there's a reason for that, that'd be interesting to mention)

Reply:
I've added a comment 👍